cargo_buildscript.bzl incompatible with latest Buck2 prelude

## Summary

The `cargo_buildscript.bzl` in buckal-bundles loads `DepCollectionContext` from `@prelude//rust:context.bzl`, but the latest Buck2 rolling binary (build `2026-03-17-c5c2e6b13c614ef5ca665facbbfd000e6d7a2b33`) no longer exports this symbol, causing all `buck2 build` invocations to fail.

## Error

```
BUILD FAILED
From load at third-party/rust/crates/serde_urlencoded/0.7.1/BUCK:4

Caused by:
    0: From load at buckal/wrapper.bzl:6
    1: Error evaluating module: `buckal//cargo_buildscript.bzl@root`
    2: error: Module symbol `DepCollectionContext` is not exported
         --> buckal/cargo_buildscript.bzl:42:36
          |
       42 | load("@prelude//rust:context.bzl", "DepCollectionContext")
          |                                    ^^^^^^^^^^^^^^^^^^^^^^
```

## Impact

- All integration tests (fd, git-internal, libra, monorepo-demo) fail on all platforms (Linux, macOS, Windows)
- Any user running `cargo buckal build` with the latest Buck2 binary will hit this error

## Workaround

Pin Buck2 to `2026-02-01` (done in cargo-buckal CI via buck2hub/cargo-buckal@c9f817f).

## Fix needed

Update `cargo_buildscript.bzl` to be compatible with the current Buck2 prelude (remove or replace the `DepCollectionContext` import).

--- a/cargo_buildscript.bzl
+++ b/cargo_buildscript.bzl
@@ -24,6 +24,7 @@ load("@prelude//decls:toolchains_common.bzl", "toolchains_common")
 load("@prelude//os_lookup:defs.bzl", "Os", "OsLookup")
 load("@prelude//rust:rust_toolchain.bzl", "RustToolchainInfo")
 load("@prelude//rust:targets.bzl", "targets")
+load("@prelude//rust/tools:attrs.bzl", "RustInternalToolsInfo")
 load(
     "@prelude//rust/tools:buildscript_platform.bzl",
     "buildscript_platform_constraints",
@@ -39,7 +40,7 @@ load(
     "get_reindeer_platform_names",
     "get_reindeer_platforms",
 )
-load("@prelude//rust:context.bzl", "DepCollectionContext")
+load("@prelude//rust:dep_context.bzl", "DepCollectionContext")
 load(
     "@prelude//rust:link_info.bzl",
     "DEFAULT_STATIC_LINK_STRATEGY",
@@ -65,21 +66,27 @@ def _make_rustc_shim(ctx: AnalysisContext, cwd: Artifact) -> cmd_args:
         )
         deps = gather_explicit_sysroot_deps(dep_ctx)
         deps = resolve_rust_deps_inner(ctx, deps)
-        dep_args, _ = dependency_args(
+        dep_args, dep_argsfiles, _ = dependency_args(
             ctx = ctx,
-            compile_ctx = None,
+            internal_tools_info = ctx.attrs._rust_internal_tools_toolchain[RustInternalToolsInfo],
+            transitive_dependency_dirs = set(),
             toolchain_info = toolchain_info,
             deps = deps,
             subdir = "any",
             dep_link_strategy = DEFAULT_STATIC_LINK_STRATEGY,
             dep_metadata_kind = MetadataKind("full"),
             is_rustdoc_test = False,
+            cwd = cwd,
         )
 
         null_path = "nul" if ctx.attrs._exec_os_type[OsLookup].os == Os("windows") else "/dev/null"
         dep_args = cmd_args("--sysroot=" + null_path, dep_args, relative_to = cwd)
         dep_file, _ = ctx.actions.write("rustc_dep_file", dep_args, allow_args = True)
-        sysroot_args = cmd_args("@", dep_file, delimiter = "", hidden = dep_args)
+        sysroot_args = cmd_args(
+            cmd_args("@", dep_file, delimiter = "", hidden = dep_args),
+            # add dep_argsfiles as a separate argument because rustc does NOT support nested @argsfiles
+            dep_argsfiles,
+        )
     else:
         sysroot_args = cmd_args()
 
@@ -105,7 +112,7 @@ def _cargo_buildscript_impl(ctx: AnalysisContext) -> list[Provider]:
         manifest_dir = ctx.attrs.manifest_dir[DefaultInfo].default_outputs[0]
     else:
         manifest_dir = ctx.actions.symlinked_dir("manifest_dir", ctx.attrs.filegroup_for_manifest_dir)
-        
+
     cmd = [
         ctx.attrs.runner[RunInfo],
         cmd_args("--buildscript=", ctx.attrs.buildscript[RunInfo], delimiter = ""),
@@ -154,6 +161,14 @@ def _cargo_buildscript_impl(ctx: AnalysisContext) -> list[Provider]:
         upper_feature = feature.upper().replace("-", "_")
         env["CARGO_FEATURE_{}".format(upper_feature)] = "1"
 
+    for flag in toolchain_info.rustc_flags:
+        if isinstance(flag, ResolvedStringWithMacros):
+            flag = str(flag)[1:-1]
+        if flag.startswith("-Copt-level="):
+            opt_level = flag.removeprefix("-Copt-level=")
+            if opt_level.isdigit():
+                env["OPT_LEVEL"] = opt_level
+
     # Environment variables specified in the target's attributes get priority
     # over all the above.
     for k, v in ctx.attrs.env.items():
@@ -192,6 +207,9 @@ _cargo_buildscript_rule = rule(
         "rustc_host_tuple": attrs.dep(default = "prelude//rust/tools:rustc_host_tuple"),
         "version": attrs.string(),
         "_exec_os_type": buck.exec_os_type_arg(),
+        "_rust_internal_tools_toolchain": attrs.default_only(
+            attrs.toolchain_dep(default = "prelude//rust/tools:internal_tools_toolchain"),
+        ),
         "_rust_toolchain": toolchains_common.rust(),
     },
     # Always empty, but needed to prevent errors

